File: binary-tree-postorder-traversal.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <deque>
#include <unordered_map>
#include <list>
#include <stack>

using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {

stack<TreeNode*> Stack;

public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        // order: mid, right, left
        TreeNode *ptr = root; // node not yet output or pushed to stack
        while (!Stack.empty() || ptr) {
            if (ptr) {
                ans.push_back(ptr->val); // each node can only be visited once
                Stack.push(ptr->left); // all left children will be popped after mid/right visited
                ptr = ptr->right;
            } else {
                ptr = Stack.top(); // a left child
                Stack.pop();
            }
        }

        reverse(ans.begin(), ans.end());
        return ans;
    }
};

int main() {
    Solution solution;

    return 0;
}

File: lintcode/binary-tree-postorder-traversal.java
// https://www.jiuzhang.com/solution/binary-tree-postorder-traversal/#tag-highlight-lang-cpp
// 81.8%, 4.8%

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {

    List<TreeNode> stack;

    private TreeNode stackTop() {
        return stack.get(stack.size()-1);
    }

    private void stackPop() {
        stack.remove(stack.size()-1);
    }

    /**
     * @param root: A Tree
     * @return: Postorder in ArrayList which contains node values.
     */
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList();
        if (root == null) {
            return ans;
        }
        TreeNode p = null;
        stack = new ArrayList<>();
        stack.add(root);
        while (!stack.isEmpty()) {
            TreeNode node = stackTop();
            if (hasVisited(node, p)) {
                ans.add(node.val);
                p = node;
                stackPop();
            } else {
                if (node.right != null) {
                    stack.add(node.right);
                }
                if (node.left != null) {
                    stack.add(node.left);
                }
            }
        }
        return ans;
    }

    private boolean hasVisited(TreeNode node, TreeNode prev) {
        if (node.right != null) {
            return (node.right == prev);
        } else {
            if (node.left != null) {
                return (node.left == prev);
            } else {
                return true;
            }
        }
    }

}
